### ISSUE TITLE
[BUG] Callbacks fail to regenerate on second webapp testing instances.

### ISSUE THREAD

**Issue by Andre-Medina** [plotly/dash#3024]
**Describe your context**
Please provide us your environment, so we can easily reproduce the issue.

```
Package                               Version                         Editable project location
------------------------------------- ------------------------------- ----------------------------------------------------------
absl-py                               2.1.0
aiohappyeyeballs                      2.4.3
aiohttp                               3.10.8
aiosignal                             1.3.1
amply                                 0.1.6
amqp                                  5.2.0
annotated-types                       0.7.0
anyio                                 4.6.0
argon2-cffi                           23.1.0
argon2-cffi-bindings                  21.2.0
arrow                                 1.3.0
arviz                                 0.20.0
asciitree                             0.3.3
asttokens                             2.4.1
astunparse                            1.6.3
async-lru                             2.0.4
async-timeout                         4.0.3
attrs                                 24.2.0
Authlib                               1.3.0
Babel                                 2.14.0
backoff                               2.2.1
backports.tarfile                     1.0.0
bcrypt                                4.2.0
beautifulsoup4                        4.12.3
billiard                              4.2.1
bleach                                6.1.0
blinker                               1.8.2
bokeh                                 3.5.2
boto3                                 1.35.31
botocore                              1.35.31
branca                                0.7.2
Brotli                                1.1.0
CacheControl                          0.14.0
cached-property                       1.5.2
cachelib                              0.9.0
cachetools                            5.5.0
celery                                5.4.0
certifi                               2024.8.30
cffi                                  1.17.1
cfgv                                  3.3.1
charset-normalizer                    3.3.2
click                                 8.1.7
click-didyoumean                      0.3.1
click-plugins                         1.1.1
click-repl                            0.3.0
cligj                                 0.7.2
cloud-store                           1.2.8
cloudpickle                           3.0.0
coiled                                1.53.0
colorama                              0.4.6
comm                                  0.2.2
cons                                  0.4.6
contourpy                             1.3.0
cryptography                          43.0.1
cycler                                0.12.1
cytoolz                               0.12.3
dash                                  2.18.1
dash_ag_grid                          31.2.0
dash_auth                             2.3.0
dash-breakpoints                      0.1.0
dash-core-components                  2.0.0
dash_cytoscape                        1.0.2
dash-deck                             0.0.1
dash-extensions                       1.0.18
dash-html-components                  2.0.0
dash_iconify                          0.1.2
dash-intersection-observer            1.0.1
dash_leaflet                          1.0.15
dash_mantine_components               0.12.1
dash-parallel-callback                0.6.3
dash-table                            5.0.0
dash-testing-stub                     0.0.2
dask                                  2024.5.2
dask-cluster                          0.3.5
dask-expr                             1.1.2
dask-gateway                          2024.1.0
dataclass-wizard                      0.22.3
debugpy                               1.8.6
decorator                             5.1.1
defusedxml                            0.7.1
Deprecated                            1.2.14
der-core                              0.6.3
dill                                  0.3.9
distlib                               0.3.8
distributed                           2024.5.2
dm-tree                               0.1.8
docutils                              0.21.2
EditorConfig                          0.12.4
enea-appshell                         0.9.4
entrypoints                           0.4
ephem                                 4.1.5
et-xmlfile                            1.1.0
etuples                               0.3.9
exceptiongroup                        1.2.2
executing                             2.1.0
fabric                                3.2.2
Faker                                 30.1.0
fasteners                             0.17.3
fastjsonschema                        2.20.0
filelock                              3.16.1
fiona                                 1.9.6
firebase-admin                        6.5.0
firestore-aio                         1.12.1
Flask                                 2.2.5
Flask-Caching                         2.3.0
Flask-Compress                        1.15
flatbuffers                           24.3.25
flatdict                              4.0.1
folium                                0.17.0
fonttools                             4.54.1
fqdn                                  1.5.1
frozenlist                            1.4.1
fsspec                                2024.9.0
gast                                  0.5.5
gcsfs                                 2024.9.0.post1
GDAL                                  3.8.4
geobuf                                1.1.1
geopandas                             1.0.1
gilknocker                            0.4.1
gitdb                                 4.0.11
GitPython                             3.1.43
google-api-core                       2.20.0
google-api-python-client              2.147.0
google-auth                           2.35.0
google-auth-httplib2                  0.2.0
google-auth-oauthlib                  1.2.1
google-cloud-appengine-logging        1.3.0
google-cloud-audit-log                0.3.0
google-cloud-core                     2.4.1
google-cloud-firestore                2.19.0
google-cloud-logging                  3.11.2
google-cloud-secret-manager           2.20.1
google-cloud-storage                  2.18.2
google-cloud-tasks                    2.16.5
google-cloud-workflows                1.10.1
google-crc32c                         1.1.2
google-pasta                          0.2.0
google-resumable-media                2.7.2
googleapis-common-protos              1.65.0
gprof2dot                             2024.6.6
grpc-google-iam-v1                    0.13.1
grpcio                                1.59.3
grpcio-status                         1.59.3
gunicorn                              23.0.0
h11                                   0.14.0
h2                                    4.1.0
h5netcdf                              1.3.0
h5py                                  3.11.0
holidays                              0.57
hpack                                 4.0.0
httpcore                              1.0.6
httplib2                              0.22.0
httpx                                 0.27.2
humanize                              4.10.0
hyperframe                            6.0.1
identify                              2.6.1
idna                                  3.10
importlib_metadata                    7.0.2
importlib_resources                   6.4.5
iniconfig                             2.0.0
invoke                                2.2.0
ipykernel                             6.29.5
ipython                               8.27.0
ipywidgets                            8.1.5
isoduration                           20.11.0
itsdangerous                          2.2.0
jaraco.classes                        3.4.0
jaraco.context                        5.3.0
jaraco.functools                      4.0.0
jax                                   0.4.27
jaxlib                                0.4.23.dev20240214
jedi                                  0.19.1
jeepney                               0.8.0
Jinja2                                3.1.4
jmespath                              1.0.1
joblib                                1.4.2
jsbeautifier                          1.15.1
json5                                 0.9.25
jsondiff                              2.0.0
jsonpointer                           3.0.0
jsonschema                            4.23.0
jsonschema-specifications             2023.12.1
jupyter_client                        8.6.3
jupyter_core                          5.7.2
jupyter-events                        0.10.0
jupyter-lsp                           2.2.5
jupyter_server                        2.14.2
jupyter_server_terminals              0.5.3
jupyterlab                            4.2.5
jupyterlab_pygments                   0.3.0
jupyterlab_server                     2.27.3
jupyterlab_widgets                    3.0.13
keras                                 2.15.0
keyring                               25.4.1
keyrings.google-artifactregistry-auth 1.1.1
kiwisolver                            1.4.7
kombu                                 5.4.1
lazy-import                           0.2.2
lazy_loader                           0.4
llvmlite                              0.43.0
locket                                1.0.0
logical-unification                   0.4.6
lxml                                  5.3.0
lz4                                   4.3.3
mapclassify                           2.8.1
Markdown                              3.6
markdown-it-py                        3.0.0
MarkupSafe                            2.1.5
matplotlib                            3.9.2
matplotlib-inline                     0.1.7
mdurl                                 0.1.2
miniKanren                            1.0.3
mistune                               3.0.2
ml-dtypes                             0.2.0
more-itertools                        10.5.0
msgpack                               1.1.0
msgspec                               0.18.6
multidict                             6.1.0
multimethod                           1.9.1
multipledispatch                      0.6.0
multiprocess                          0.70.16
munkres                               1.1.4
mypy-extensions                       1.0.0
nbclient                              0.10.0
nbconvert                             7.16.4
nbformat                              5.10.4
nest_asyncio                          1.6.0
networkx                              3.3
nodeenv                               1.9.1
notebook_shim                         0.2.4
numba                                 0.60.0
numcodecs                             0.13.0
numpy                                 1.26.0
numpy-financial                       1.0.0
nutpie                                0.13.1
oauthlib                              3.2.2
openpyxl                              3.1.5
opentelemetry-api                     1.27.0
opt_einsum                            3.4.0
orjson                                3.10.7
outcome                               1.3.0.post0
overrides                             7.7.0
packaging                             24.1
pandas                                2.2.3
pandera                               0.20.4
pandocfilters                         1.5.0
paramiko                              3.5.0
parso                                 0.8.4
partd                                 1.4.2
patsy                                 0.5.6
percy                                 2.0.2
pexpect                               4.9.0
pickleshare                           0.7.5
pillow                                10.3.0
pip                                   24.2
pip-requirements-parser               32.0.1
pixi-kernel                           0.4.0
pkgutil_resolve_name                  1.3.10
planning                              1.2.1a3
platformdirs                          4.3.6
plotly                                5.24.1
plotly-resampler                      0.10.0
pluggy                                1.5.0
polars                                1.2.1
polyfactory                           2.17.0
pre_commit                            3.8.0
prometheus_client                     0.21.0
prompt_toolkit                        3.0.48
proto-plus                            1.23.0
protobuf                              4.24.4
psutil                                6.0.0
ptyprocess                            0.7.0
PuLP                                  2.8.0
pure_eval                             0.2.3
pvlib                                 0.11.1
pyarrow                               15.0.0
pyarrow-hotfix                        0.6
pyasn1                                0.6.1
pyasn1_modules                        0.4.1
pycparser                             2.22
pydantic                              2.9.2
pydantic_core                         2.23.4
Pygments                              2.18.0
PyJWT                                 2.9.0
pymc                                  5.16.2
PyNaCl                                1.5.0
pyogrio                               0.7.2
pyOpenSSL                             24.2.1
pyparsing                             3.1.4
pyproj                                3.6.1
PySocks                               1.7.1
pysolar                               0.11
pytensor                              2.25.4
pytest                                8.3.3
pytest-mock                           3.14.0
pytest-profiling                      1.7.0
pytest-subtests                       0.13.1
python-dateutil                       2.9.0
python-dotenv                         1.0.1
python-json-logger                    2.0.7
python-slugify                        8.0.4
pytz                                  2024.1
pyu2f                                 0.1.5
pyxlsb                                1.0.10
PyYAML                                6.0.2
pyzmq                                 26.2.0
referencing                           0.35.1
requests                              2.32.3
requests-oauthlib                     2.0.0
retrying                              1.3.3
rfc3339-validator                     0.1.4
rfc3986-validator                     0.1.1
rich                                  13.9.1
rpds-py                               0.20.0
rsa                                   4.9
ruff                                  0.4.10
runtype                               0.5.1
ruptures                              1.1.9
s3transfer                            0.10.2
scikit-learn                          1.5.2
scipy                                 1.14.1
SecretStorage                         3.3.3
selenium                              4.2.0
Send2Trash                            1.8.3
setuptools                            75.1.0
shapely                               2.0.4
shellingham                           1.5.4
six                                   1.16.0
smmap                                 5.0.0
sniffio                               1.3.1
sortedcontainers                      2.4.0
soupsieve                             2.5
stack-data                            0.6.2
statsmodels                           0.14.3
tblib                                 3.0.0
tenacity                              9.0.0
tensorboard                           2.15.2
tensorboard-data-server               0.7.0
tensorflow                            2.15.0
tensorflow_estimator                  2.15.0
tensorflow-probability                0.23.0
termcolor                             2.4.0
terminado                             0.18.1
text-unidecode                        1.3
tf_keras                              2.15.0
threadpoolctl                         3.5.0
tinycss2                              1.3.0
toml                                  0.10.2
tomli                                 2.0.1
toolz                                 0.12.1
tornado                               6.4.1
tqdm                                  4.66.5
trace-updater                         0.0.9.1
traitlets                             5.14.3
trio                                  0.26.2
trio-websocket                        0.11.1
tsdownsample                          0.1.3
typeguard                             4.3.0
typer                                 0.12.5
typer-slim                            0.12.5
types-python-dateutil                 2.9.0.20240906
typing_extensions                     4.12.2
typing-inspect                        0.9.0
typing-utils                          0.1.0
tzdata                                2024.2
ukkonen                               1.0.1
unicodedata2                          15.1.0
Unidecode                             1.3.8
uri-template                          1.3.0
uritemplate                           4.1.1
urllib3                               2.2.3
vine                                  5.1.0
virtualenv                            20.26.6
wcwidth                               0.2.13
webcolors                             24.8.0
webencodings                          0.5.1
websocket-client                      1.8.0
Werkzeug                              2.2.3
wheel                                 0.44.0
widgetsnbextension                    4.0.13
windpowerlib                          0.2.1
wrapt                                 1.14.1
wsproto                               1.2.0
xarray                                2024.9.0
xarray-einstats                       0.8.0
xlrd                                  2.0.1
xlsx2csv                              0.8.3
XlsxWriter                            3.2.0
xyzservices                           2024.9.0
yarl                                  1.13.1
zarr                                  2.18.3
zict                                  3.0.0
zipp                                  3.20.2
zstandard                             0.23.0
```
-  If frontend related, tell us your Browser, Version and OS

    - OS: Windows
    - Browser: Chrome
    - Version: ChromeDriver 128.0.6613.137
    
**Describe the bug**

When creating pytests for a full webapp with several callbacks and pages, using `dash_duo` in the `dash[testing]` package. The first pytest will work as expected. However, something to do with the callbacks failing to activate/ regenerate on the second instance of `dash_duo` causes some callbacks to fail to trigger. Leading to the second pytest failing.

this causes very odd behavouir, where testing multiple pytests on their own will succeed. However, testing them together will cause them to fail. Very hard to debug.

**Expected behavior**

For the consecutive tests to run as if the first test did not.

The following code form @RenaudLN fixes the issue:

```python
from copy import deepcopy

import dash._callback
import pytest

PYDF_CALLBACK_LIST = []
PYDF_CALLBACK_MAP = {}


@pytest.fixture(scope="session", autouse=True)
def init_test_session():
    """Store all registered callbacks."""
    global PYDF_CALLBACK_LIST, PYDF_CALLBACK_MAP  # noqa: PLW0603
    PYDF_CALLBACK_LIST = deepcopy(dash._callback.GLOBAL_CALLBACK_LIST)
    PYDF_CALLBACK_MAP = deepcopy(dash._callback.GLOBAL_CALLBACK_MAP)


@pytest.fixture(scope="function", autouse=True)
def reset_callbacks():
    """Add global callbacks back."""
    dash._callback.GLOBAL_CALLBACK_LIST = deepcopy(PYDF_CALLBACK_LIST)
    dash._callback.GLOBAL_CALLBACK_MAP = deepcopy(PYDF_CALLBACK_MAP)
```

**Screenshots**

**Comment by T4rk1n** [comment]
This is by design, the callbacks are only generated for the apps, they are flushed for being able to define multiple apps with different callbacks in the testing.

There are several patterns that can workaround this, use an app factory function to create the callbacks at the same time as the app. You can also have a `session` fixture with your app instead and only start the app/driver once and resetting the url between tests.

In Dash 3.0, there will be a new `hooks` system that will be able to add callbacks to all the apps.

**Comment by Andre-Medina** [comment]
okay thanks!

After investigating, the above solution didnt complete fix the issue. found it was due to having a single fixture that took `dash_duo` and generated an app within this fixture, to then start with `dash_duo.start_server(app, host="localhost")`.

as this fixture the `pytest.fixture` `scope` of `"function"`. The callbacks were flushed after each function. 

To fix this further, along with the fixtures above, I separated the fixture that created the `app` to its own `pytest.fixture` with the `scope` of `session` and `autouse=True`, then passed this `app` fixture into the `dash_duo` fixture.
